add(book): Describes custom Testnets in the Zebra book (#8636)
ZcashFoundation/zebra: book/src/user/custom-testnets.md

- Slow start intervals, 
- Genesis hashes, and
- Target difficulty limits.

It's also possible to disable Proof-of-Work validation by setting `disable_pow` to `true` so that blocks can be mined onto the chain without valid Equihash solutions, nor block hashes below their target difficulties.

All of these parameters are optional, if they are all omitted or set to their default values, Zebra will run on the default public Testnet.

## Usage

In order to use a custom Testnet, Zebra must be configured to run on Testnet with non-default Testnet parameters. If the node is meant to mine blocks, it will need a `[mining]` section, and if it's meant to mine blocks with non-coinbase transactions, it will also need the `[rpc]` section so the `send_raw_transaction` RPC method is available.

Relevant parts of the configuration file:

```toml
[mining]
miner_address = 't27eWDgjFYJGVXmzrXeVjnb5J3uXDM9xH9v'
    
[network]
network = "Testnet"

[network.testnet_parameters.activation_heights]
network_name = "ConfiguredTestnet_1"
network_magic = [0, 1, 0, 255]
slow_start_interval = 0
target_difficulty_limit = "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f"
disable_pow = true

> # Configured activation heights must be greater than or equal to 1, and less than

Reviewer: ```suggestion
# Configured activation heights must be greater than 0, and less than
```

# or equal to Zebra's max block height of 2^31 - 1, block height 0 is reserved

> # for the Genesis network upgrade in Zebra.

Reviewer: Zebra's max block height in the state is actually 16777215: https://github.com/zcashfoundation/zebra/blob/d1a9004b3290add477f9a83ce959b7538725be1d/zebra-state/src/service/finalized_state/disk_format/block.rs#L35, but I don't think it applies here.

Maybe let's go with the following:

```suggestion
# 2^31. Block height 0 is reserved for the Genesis network upgrade in Zebra.
```

Author: > Zebra's max block height in the state is actually 16777215

That's a good point, I'll add a footnote about the max on-disk serialized block height currently being `2^24 - 1`.

Update: Added in https://github.com/ZcashFoundation/zebra/pull/8636/commits/14147916707645cfee7c653abb569a1dbb5f871c.

#
# Network upgrades must be activated in the order that they were added to Zcash,
# configuring the activation heights of recent network upgrades will activate
# any omitted prior network upgrades at the same height. 
#
# For example, configuring the activation height of NU5 to block height 1 without
# configuring any other network upgrade activation heights will set the
# activation heights of BeforeOverwinter, Overwinter, Sapling, Blossom, 
# Heartwood, and Canopy at block height 1 as well.
[network.testnet_parameters.activation_heights]
NU5 = 1

# This section may be omitted if it's not necessary to 
# add transactions to Zebra's mempool
[rpc]
listen_addr = "0.0.0.0:18232"
```

> 

Reviewer: ```suggestion
```

> Relevant parts of the configuration file with Mainnet parameters:

Reviewer: We mention here that the parameters correspond to Mainnet, but the config seems to be for Testnet.

```suggestion
Relevant parts of the configuration file with Testnet parameters:
```

I think it's better to go with Testnet because the mining address below is for Testnet.

Author: It is a Testnet, only the activation heights, slow start interval, genesis hash, and target difficulty limit match the Mainnet parameters, because the other ones are reserved. It's mentioned in the "Configuring More Testnet Parameters To Match Mainnet Consensus Rules" section that the Mainnet address prefixes and coin type should be reserved for only Mainnet, and it's mentioned in "Caveats and Restrictions" that the Mainnet network name and magic are reserved as well.

I'll list those parameters explicitly here so it's clear that the network name, network magic, address prefixes, and coin type still do not match Mainnet.

> I think it's better to go with Testnet because the mining address below is for Testnet

The first config has more practical parameters for a custom Testnet, this one is meant to be an example of how to configure a custom Testnet to more closely match Mainnet parameters. I'll note above that the first config is more practical.

Added "with practical parameters" to the heading before the first config and a footnote about what parameters in the second config actually match Mainnet in https://github.com/ZcashFoundation/zebra/pull/8636/commits/14147916707645cfee7c653abb569a1dbb5f871c.

```toml
[mining]
miner_address = 't27eWDgjFYJGVXmzrXeVjnb5J3uXDM9xH9v'
    
[network]
network = "Testnet"

[network.testnet_parameters.activation_heights]
# The Mainnet, Testnet, and Regtest network names are reserved.
network_name = "ConfiguredTestnet_2"

> # The Mainnet and Regtest network magics are reserved.

Reviewer: Should we mention Testnet as well?

```suggestion
# The Mainnet, Testnet and Regtest network magics are reserved.
```

Author: The Testnet network magic actually isn't reserved, though it may be a good idea to only allow it for compatible custom Testnets (currently configuring anything except the network name will result in an incompatible Testnet).

Update: Added a note about not recommending usage of the default Testnet network magic with incompatible parameters in https://github.com/ZcashFoundation/zebra/pull/8636/commits/14147916707645cfee7c653abb569a1dbb5f871c.

> network_magic = [0, 1, 0, 255]

Reviewer: We mention above that the parameters correspond to Mainnet, but the magic doesn't seem like any of the ones we have here https://github.com/zcashfoundation/zebra/blob/d1a9004b3290add477f9a83ce959b7538725be1d/zebra-chain/src/parameters/constants.rs#L24.

Shouldn't the magic be like this for Testnet instead?

```suggestion
network_magic = [250, 26, 249, 191]
```

Or we can't use the one for Testnet because we're using the default initial peers?

Author: > Or we can't use the one for Testnet because we're using the default initial peers?

We could use the default Testnet magic here, but can't use the default initial peers because the parameters in both of these configs are incompatible with the default Testnet. I don't think we should use the default Testnet magic because the other parameters are incompatible, but it shouldn't make much difference in practice.

Reviewer: I see. Thanks for the clarification. Let's not use the Testnet magic.

slow_start_interval = 20_000
genesis_hash = "00040fe8ec8471911baa1db1266ea15dd06b4a8a5c453883c000b031973dce08"

# Note that setting `disable_pow` to `false` with this target difficultly 
# limit will make it very difficult to mine valid blocks onto the chain.
target_difficulty_limit = "0008000000000000000000000000000000000000000000000000000000000000"
disable_pow = false

[network.testnet_parameters.activation_heights]
BeforeOverwinter = 1
Overwinter = 347_500
Sapling = 419_200
Blossom = 653_600
Heartwood = 903_000
Canopy = 1_046_400

> NU5 = 1_687_104

Reviewer: These are for Mainnet, so if we apply the other suggestions, we should change these so they match Testnet.

Author: I'd like to keep these so that it's possible to copy the Mainnet best chain and checkpoints over to a custom Testnet with this config and disable the Proof-of-Work check to add blocks to it.

Added a comment above this config field about when to use these activation heights in https://github.com/ZcashFoundation/zebra/pull/8636/commits/14147916707645cfee7c653abb569a1dbb5f871c.

# This section may be omitted if it's not necessary to 
# add transactions to Zebra's mempool
[rpc]
listen_addr = "0.0.0.0:18232"
```

## Caveats and Restrictions

There are a few caveats:

> - Configured network upgrade activation heights must be above the genesis block height, which is reserved for Zebra's `Genesis` network upgrade, and must not be above Zebra's max block height of `2^31 -1`.

Reviewer: ```suggestion
- Configured network upgrade activation heights must be above the genesis block height, which is reserved for Zebra's `Genesis` network upgrade, and must not be above Zebra's max block height of `2^31 - 1`.
```

- While it's possible to activate Canopy and later network upgrades after height 1, Zebra cannot currently produce pre-Canopy block templates, so the `getblocktemplate` RPC method and Zebra's internal miner which depends on the `getblocktemplate` method won't work until Canopy is activated. An alternative block source will be required to mine pre-Canopy blocks onto Zebra's chain.

> - While it's possible to use the default Testnet network magic with a configured Testnet, Zebra will panic when configured to use the default initial Testnet peers and Testnet parameters that are incompatible with the default public Testnet.

Reviewer: > ... Zebra will panic when configured to use the [...] Testnet parameters that are incompatible with the default public Testnet.

Should we document what those are?

Author: > Should we document what those are?

Yes! Added a footnote and a line in the introduction about which fields will result in an incompatible Testnet in https://github.com/ZcashFoundation/zebra/pull/8636/commits/14147916707645cfee7c653abb569a1dbb5f871c.

- If the genesis hash is configured, a genesis block will need to be copied into the custom Testnet state or submitted via the `submitblock` RPC method, Zebra cannot currently generate genesis blocks. See the `CreateGenesisBlock()` function in `zcashd/src/chainparams.cpp` for use cases that require a new genesis block.

There are also a few other restrictions on these parameters:
- The network name must:
    - not be any of the reserved network names: `Testnet`, `Mainnet`, and `Regtest`,
    - contain only alphanumeric characters and underscores, and
    - be shorter than the `MAX_NETWORK_NAME_LENGTH` of `30`.

> - The network magic must not be any of the reserve network magics: `[36, 233, 39, 100]` and `[170, 232, 63, 95]`, these are the `Mainnet` and `Regtest` network magics respectively.

Reviewer: ```suggestion
- The network magic must not be any of the reserved network magics: `[36, 233, 39, 100]` and `[170, 232, 63, 95]`, these are the `Mainnet` and `Regtest` network magics respectively.
```

- The network upgrade activation heights must be in order, such that the activation height for every network upgrade is at or above the activation height of every preceding network upgrade.

## Comparison To Mainnet and Default Public Testnet Consensus Rules

> Aside from the configurable parameters, custom Testnets in Zebra validate all of same consensus rules as Testnet.

Reviewer: ```suggestion
Aside from the configurable parameters, custom Testnets in Zebra validate the same consensus rules as Testnet.
```

- Max block time is always enforced on Mainnet, but only enforced after block height `653,606` on Testnets (block times later than the median block time + a constant are rejected while this rule is enforced, this is part of the block header consensus rules in the Zcash specification).

### Configuring More Testnet Parameters To Match Mainnet Consensus Rules

The Mainnet Zcash address prefixes and coin type should remain reserved for Mainnet.

The network upgrade activation heights and target difficulty limit of a custom Testnet could currently be configured in Zebra to match Mainnet.

The remaining consensus differences between Mainnet and Testnet could be made configurable in Zebra so that they could be configured to match Mainnet.

## Differences Between Custom Testnets and Regtest

Zebra's Regtest network is a special case of a custom Testnet that:
- Won't make peer connections[^fn1],
- Skips Proof-of-Work validation,
- Uses a reserved network magic and network name,
- Activates network upgrades up to and including Canopy at block height 1,
- Tries to closely match the `zcashd` Regtest parameters, and
- Expects the Regtest genesis hash.

> Once Zebra's internal miner can mine Equihash solutions configurable parameters, Regtest should validate Proof-of-Work with the zcashd Equihash parameters unless it's disabled in its configuration, and custom Testnets should allow for configuring their Equihash parameters as well.

Reviewer: ```suggestion
Once Zebra's internal miner can mine Equihash solutions with configurable parameters, Zebra's Regtest should validate Proof-of-Work with the zcashd Regtest Equihash parameters unless it's disabled in its configuration, and custom Testnets should allow for configuring their Equihash parameters as well.
```

# Custom Testnets

Custom Testnets in Zebra enable testing consensus rule changes on a public, configured Testnet independent of the default public Zcash Testnet.

Zebra's Testnet can be configured with custom:
- Network upgrade activation heights, 
- Network names, 
- Network magics, 
- Slow start intervals, 
- Genesis hashes, and
- Target difficulty limits.

> Configuring any of those Testnet parameters except the network name with non-default values will result in an incompatible custom Testnet.
> 
> It's also possible to disable Proof-of-Work validation by setting `disable_pow` to `true` so that blocks can be mined onto the chain without valid Equihash solutions, nor block hashes below their target difficulties.

Author: ```suggestion
It's also possible to disable Proof-of-Work validation by setting `disable_pow` to `true` so that blocks can be mined onto the chain without valid Equihash solutions, nor block hashes below their target difficulties.

Configuring any of those Testnet parameters except the network name with non-default values will result in an incompatible custom Testnet.
```

> Configuring any of those Testnet parameters except the network name with non-default values will result in an incompatible custom Testnet.

Reviewer: Can we also clarify what an "incompatible Testnet" means?

Reviewer: We mention it a few times later on, so I think it would be useful to have explained what we mean by the incompatibility.

Author: Described what's meant by "incompatible Testnet" in https://github.com/ZcashFoundation/zebra/pull/8636/commits/0842950c785dca8826127029a9cf26470214dfa5.

> Configuring any of those Testnet parameters except the network name with non-default values will result in an incompatible custom Testnet.

Author: ```suggestion
Configuring any of those Testnet parameters except the network name with non-default values will result in an incompatible custom Testnet. Incompatible Testnets will fail to successfully complete peer handshakes with one another, or could provide one another with invalid blocks or invalid mempool transactions. Peer node connections that consistently provide invalid blocks or mempool transactions should be considered misbehaving peer connections and dropped.
```

##### Footnotes

[^fn1]: Only the network upgrade activation heights, target difficulty limit, slow start interval, and genesis hash of a custom Testnet may currently be configured in Zebra to match Mainnet, other differences between Mainnet and Testnet consensus rules still remain with any custom Testnet configuration. More parameters may become configurable in the future, but the Mainnet Zcash address prefixes, coin type, network name, and network magic should remain reserved only for Mainnet.

[^fn2]: Zebra's max on-disk serialized block height is currently `2^24 - 1`, the max block height of `2^31 - 1` can only be represented in-memory, so while an activation height of `2^31 - 1` is valid, Zebra's best chain would not currently be able to reach that activation height.

> [^fn3]: Configuring any of the Testnet parameters that are currently configurable except the network name will result in an incompatible custom Testnet, these are: the network magic, network upgrade activation heights, slow start interval, genesis hash, and target difficulty limit.

Author: ```suggestion
[^fn3]: Configuring any of the Testnet parameters that are currently configurable except the network name will result in an incompatible custom Testnet, these are: the network magic, network upgrade activation heights, slow start interval, genesis hash, disabled Proof-of-Work and target difficulty limit.
```
